### DISCUSSION TITLE
Assigning the result of a ODE solve to function

### DISCUSSION THREAD

**Discussion by max3-2** [sympy/sympy#23310] [Q&A]
Hello,

I am trying to solve an ODE (works) and use the result as a function onward, e.g more integrations, more evaluation, more ODEs. However the result is an `Equality`, or if I use `.rhs` its an `Expression`. I can't seem to find a way to reassign this as a function. Am I overlooking something? Is my use wrong and using `subs` or `lambdify` is the way to go numerically and using `rhs` for analytical processing? I hope my example makes it more clear, I just want to have `f(t)` as a function...

```python
import sympy as sp
from sympy.abc import a, f, t, s

f = sp.Function(f)
ode = sp.Eq(
    sp.Derivative(f(t), t), -a * f(t)
    )

sol = sp.dsolve(ode, ics={f(0): s})

# sol is an equality
print(type(sol))

# sol.rhs is an expresssion
print(type(sol.rhs))

# e.g. integrating
integral = sp.integrate(sol.rhs, t)

# yields an expression again...

# --> I am trying to keep this as a function!
```

**Comment by oscarbenjamin** [comment]
I don't think I understand what you mean by "use the result as a function". Maybe this is what you want:
```python
In [4]: from sympy.abc import a, t, s

In [5]: f = Function('f')

In [6]: ode = Eq(f(t).diff(t), -a*f(t))

In [7]: sol = dsolve(ode, ics={f(0):s})

In [8]: sol
Out[8]: 
          -a⋅t
f(t) = s⋅ℯ    

In [9]: sol.rhs
Out[9]: 
   -a⋅t
s⋅ℯ    

In [10]: fsol = Lambda(t, sol.rhs)

In [11]: fsol
Out[11]: 
       -a⋅t
t ↦ s⋅ℯ    

In [12]: fsol(1)
Out[12]: 
   -a
s⋅ℯ  
```
Note that this isn't really any different from using subs:
```python
In [13]: sol.rhs.subs(t, 1)
Out[13]: 
   -a
s⋅ℯ 
```

**Comment by max3-2** [comment]
Wow thank you for the quick reply. Thats pretty much it and not...So yes now I can use `fsol` but calling something on it does not work as I would expect (but maybe my expectations are the issue, I am an engineer and not a mathematician ;) )

E.g.
```python
integrate(fsol, t)
Out[27]: 
  ⎛       -a⋅t⎞
t⋅⎝t ↦ s⋅ℯ    ⎠
```
I would expect the integration over `t` in the function

or 
```python
from sympy.abc import b

test = 3 * b
Out[17]: 3⋅b

test = test.subs(b, fsol)
Out[18]: 
  ⎛       -a⋅t⎞
3⋅⎝t ↦ s⋅ℯ    ⎠

test(1)
Traceback (most recent call last):

  File "<ipython-input-21-b5b7b83ded54>", line 1, in <module>
    test(1)

TypeError: 'Mul' object is not callable
```

**Reply by oscarbenjamin** [reply]
The `integrate` function works with expressions:
```python
In [29]: fsol
Out[29]: 
       -a⋅t
t ↦ s⋅ℯ    

In [30]: fsol(t)
Out[30]: 
   -a⋅t
s⋅ℯ    

In [31]: integrate(fsol(t), t)
Out[31]: 
⎧    -a⋅t            
⎪-s⋅ℯ                
⎪─────────  for a ≠ 0
⎨    a               
⎪                    
⎪   s⋅t     otherwise
⎩                    

In [32]: integrate(sol.rhs, t)
Out[32]: 
⎧    -a⋅t            
⎪-s⋅ℯ                
⎪─────────  for a ≠ 0
⎨    a               
⎪                    
⎪   s⋅t     otherwise
⎩   
```

**Reply by oscarbenjamin** [reply]
As I said I'm not sure why you want a "function". If you want to call integrate then `sol.rhs` should be sufficient.

**Comment by max3-2** [comment]
Again, thank you. It helps in the current version of my code flow. There is a good chance I need to re-evaluate but since I did not find any answer I took a chance and asked. At least now I can say I do better understand the different sympy **types**

**Reply by oscarbenjamin** [reply]
Okay well if you have further questions I think that the mailing list is usually better than GitHub: https://groups.google.com/g/sympy
